File: ADT/trees/662.py
# Amazon, Flipkart, VmWare

# Level Order Traversal(BFS)

from collections import deque
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        q = deque([(root, 0)])
        max_width = 0
        while q:
            n = len(q)
            f = q[0][1]
            l = q[-1][1]
            max_width = max(max_width, l - f + 1)
            for _ in range(n):
                curr, i = q.popleft()
                if curr.left:
                    q.append((curr.left, 2 * i + 1))
                if curr.right:
                    q.append((curr.right, 2 * i + 2))
        return max_width

File: ADT/trees/662.js

var widthOfBinaryTree = function (root) {
    if (!root) return 0;
    const que = [];
    que.push({ node: root, index: 0n });
    let maxWidth = 0n;

    while (que.length > 0) {
        const n = que.length;
        const f = que[0].index;
        const l = que[que.length - 1].index;
        maxWidth = maxWidth > (l - f + 1n) ? maxWidth : (l - f + 1n);

        for (let i = 0; i < n; i++) {
            const { node: curr, index: d } = que.shift();
            if (curr.left) {
                que.push({ node: curr.left, index: 2n * d + 1n });
            }
            if (curr.right) {
                que.push({ node: curr.right, index: 2n * d + 2n });
            }
        }
    }
    return Number(maxWidth);
};